api4/preference.go

```diff
@@ -0,0 +1,113 @@
+// // Copyright (c) 2017 Mattermost, Inc. All Rights Reserved.
+// // See License.txt for license information.
+
+package api4
+
+import (
+	"net/http"
+
+	l4g "github.com/alecthomas/log4go"
+	"github.com/mattermost/platform/app"
+	"github.com/mattermost/platform/model"
+	"github.com/mattermost/platform/utils"
+)
+
+func InitPreference() {
+	l4g.Debug(utils.T("api.preference.init.debug"))
+
+	BaseRoutes.Preferences.Handle("", ApiSessionRequired(getPreferences)).Methods("GET")
+	BaseRoutes.Preferences.Handle("", ApiSessionRequired(updatePreferences)).Methods("PUT")
+	BaseRoutes.Preferences.Handle("/delete", ApiSessionRequired(deletePreferences)).Methods("POST")
+	BaseRoutes.Preferences.Handle("/{category:[A-Za-z0-9_]+}", ApiSessionRequired(getPreferencesByCategory)).Methods("GET")
+	BaseRoutes.Preferences.Handle("/{category:[A-Za-z0-9_]+}/name/{preference_name:[A-Za-z0-9_]+}", ApiSessionRequired(getPreferenceByCategoryAndName)).Methods("GET")
```

Comment:
Hmmm these look good but I think I'd like to make on modification to make these endpoints a little more useful. Can we prepend each preference endpoint with a user? So `/preferences` becomes `/users/{user_id}/preferences`. This will allow other users (mainly admins/bots) to access other users' preferences.

To handle this, your handlers will need to do permission checks of permission to the user

Sorry, I should have thought about this before but it's much apparent when I see the actual implementation :)